pkg/reconciler/taskrun/resources/taskspec_test.go

```diff
@@ -286,11 +306,224 @@ func TestGetTaskData_VerificationResult(t *testing.T) {
 	}
 	if d := cmp.Diff(verificationResult, r.VerificationResult, cmpopts.EquateErrors()); d != "" {
 		t.Errorf(diff.PrintWantGot(d))
 	}
 }
+
+func TestGetStepAction(t *testing.T) {
+	tests := []struct {
+		name           string
+		tr             *v1.TaskRun
+		want           *v1.TaskSpec
+		stepActionFunc func(ctx context.Context, n string) (*v1alpha1.StepAction, *v1.RefSource, error)
+	}{{
+		name: "step-action-with-command",
+		tr: &v1.TaskRun{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "mytaskrun",
+			},
+			Spec: v1.TaskRunSpec{
+				TaskSpec: &v1.TaskSpec{
+					Steps: []v1.Step{{
+						Ref: &v1.Ref{
+							Name: "stepAction",
+						},
+						WorkingDir: "/foo",
+					}, {
+						Ref: &v1.Ref{
+							Name: "stepAction",
+						},
+						WorkingDir: "/bar",
+						Timeout:    &metav1.Duration{Duration: time.Hour},
+					}},
+				},
+			},
+		},
+		want: &v1.TaskSpec{
+			Steps: []v1.Step{{
+				Image:      "myimage",
+				Command:    []string{"ls"},
+				Args:       []string{"-lh"},
+				WorkingDir: "/foo",
+			}, {
+				Image:      "myimage",
+				Command:    []string{"ls"},
+				Args:       []string{"-lh"},
+				WorkingDir: "/bar",
+				Timeout:    &metav1.Duration{Duration: time.Hour},
+			}},
+		},
+		stepActionFunc: getStepAction,
+	}, {
+		name: "step-action-with-script",
+		tr: &v1.TaskRun{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "mytaskrun",
+			},
+			Spec: v1.TaskRunSpec{
+				TaskSpec: &v1.TaskSpec{
+					Steps: []v1.Step{{
+						Ref: &v1.Ref{
+							Name: "stepActionWithScript",
+						},
+					}},
+				},
+			},
+		},
+		want: &v1.TaskSpec{
+			Steps: []v1.Step{{
+				Image:  "myimage",
+				Script: "ls",
+			}},
+		},
+		stepActionFunc: func(ctx context.Context, n string) (*v1alpha1.StepAction, *v1.RefSource, error) {
+			stepAction := &v1alpha1.StepAction{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "stepActionWithSctipy",
+				},
+				Spec: v1alpha1.StepActionSpec{
+					Image:  "myimage",
+					Script: "ls",
+				},
+			}
+			return stepAction, nil, nil
+		},
+	}, {
+		name: "step-action-with-env",
+		tr: &v1.TaskRun{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "mytaskrun",
+			},
+			Spec: v1.TaskRunSpec{
+				TaskSpec: &v1.TaskSpec{
+					Steps: []v1.Step{{
+						Ref: &v1.Ref{
+							Name: "stepActionWithEnv",
+						},
+					}},
+				},
+			},
+		},
+		want: &v1.TaskSpec{
+			Steps: []v1.Step{{
+				Image: "myimage",
+				Env: []corev1.EnvVar{{
+					Name:  "env1",
+					Value: "value1",
+				}},
+			}},
+		},
+		stepActionFunc: func(ctx context.Context, n string) (*v1alpha1.StepAction, *v1.RefSource, error) {
+			stepAction := &v1alpha1.StepAction{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "stepActionWithEnv",
+				},
+				Spec: v1alpha1.StepActionSpec{
+					Image: "myimage",
+					Env: []corev1.EnvVar{{
+						Name:  "env1",
+						Value: "value1",
+					}},
+				},
+			}
+			return stepAction, nil, nil
+		},
+	}, {
+		name: "step-action-error",
```

Comment:
this case can go to `TestGetStepAction_Error`

Reply:
agreed

Reply:
fixed.